### ISSUE TITLE
Support for cyrillic symbols

### ISSUE THREAD

**Issue by arteam** [asciidoctor/asciidoctor-pdf#99]
Currently I can't generate a pdf file from a document written in a cyrillic language (actually Russian). 

Cyrillic symbols are not rendered and I don't get any errors during building.
Only if I write headers in Russian I get the following error:

> /var/lib/gems/1.9.1/gems/pdf-core-0.2.5/lib/pdf/core/pdf_object.rb:55: warning: regexp match /.../n against to UTF-8 string

Example:
````asciidoc
= Test document
Doc Writer <doc@example.com>
:doctype: book
:source-highlighter: coderay
:listing-caption: Listing

Простой http://asciidoc.org[AsciiDoc] документ.

== Введение

Параграф с простым списком:

[square]
* первый элемент
* второй элемент

== Introduction

A paragraph followed by a simple list with square bullets.

[square]
* item 1
* item 2
````

**Comment by mojavelinux** [comment]
This is a similar issue to the one reported by @chloerei. See #82. Currently, it's necessary to set the path to your own fonts that have the Cyrillic characters.

As for the regexp warning, that's because Ruby 2.0 or better is required to use Prawn (without warnings). Ruby 1.9 had really screwy UTF-8 support, so we don't recommend using it.

**Comment by arteam** [comment]
Thanks, Dan, for taking time to respond to this issue.

I will try to find appropriate fonts and upgrade to the Ruby 2.x.

**Comment by sohailsed** [comment]
Hi,

I Have a similar problem.

I'm working on a java project that need to convert an asciidoc to pdf and has non-Latin chars. the encoding is UTF-8 but the non Latin char will not render in the pdf like the issue posted before.

How could I force to render non-Latin word. is there any options to do it from java(not command line) or any other solution that should be take in  consider to fix it.

I'm using the following dependencies in my project :

       <dependency>
            <groupId>org.asciidoctor</groupId>
            <artifactId>asciidoctorj</artifactId>
            <version>1.5.2</version>
        </dependency>
        <dependency>
            <groupId>org.jruby</groupId>
            <artifactId>jruby-complete</artifactId>
            <version>1.7.16.1</version>
        </dependency>
        <dependency>
            <groupId>org.asciidoctor</groupId>
            <artifactId>asciidoctorj-pdf</artifactId>
            <version>1.5.0-alpha.6</version>
        </dependency>

thanks

**Comment by sohailsed** [comment]
Hi,

I Have a similar problem.

I'm working on a java project that need to convert an asciidoc to pdf and has non-Latin chars. the encoding is UTF-8 but the non Latin char will not render in the pdf like the issue posted before.

How could I force to render non-Latin word. is there any options to do it from java(not command line) or any other solution that should be take in  consider to fix it.

I'm using the following dependencies in my project :

       <dependency>
            <groupId>org.asciidoctor</groupId>
            <artifactId>asciidoctorj</artifactId>
            <version>1.5.2</version>
        </dependency>
        <dependency>
            <groupId>org.jruby</groupId>
            <artifactId>jruby-complete</artifactId>
            <version>1.7.16.1</version>
        </dependency>
        <dependency>
            <groupId>org.asciidoctor</groupId>
            <artifactId>asciidoctorj-pdf</artifactId>
            <version>1.5.0-alpha.6</version>
        </dependency>

thanks

**Comment by mojavelinux** [comment]
At the moment, you need to create a custom theme and instruct it to use your own fonts. We have a guide about this here:

http://gist.asciidoctor.org/?github-asciidoctor%2Fasciidoctor-pdf%2F%2Fdocs%2Ftheming-guide.adoc#fonts

The reason these characters aren't supported out of the box is because I wanted to reduce the size of the gem. Fonts tend to be huge files. I trimmed down the fonts we currently bundle by subsetting the font to only include latin characters. The plan is probably to find a font that is small even with wide support (maybe NotoSerif) and ship that. Otherwise, we can just make the documentation good :)

**Comment by mojavelinux** [comment]
> /var/lib/gems/1.9.1/gems/pdf-core-0.2.5/lib/pdf/core/pdf_object.rb:55: warning: regexp match /.../n against to UTF-8 string

This warning is now fix. It was a bug in how I was adding metadata to the PDF.

**Comment by arteam** [comment]
Sorry for a late reply, but the hint from Dan worked for me. I was able to render a document in Russian
with Ruby 2.2 and Roboto font.

**Comment by mojavelinux** [comment]
Yeah!! \o/

**Comment by sohailsed** [comment]
Hi,

First of all, thank for your quick reply.

As I found, I need to define a new theme in a YAML file and put 4 fonts of a famliy somewhere in the project and use the syntax like 
 base:
  font_family: MyFont

where MyFont is my non-latin font.

Ok,good. But how could I Apply these YAML file and the fonts to my project?

in the url you've posted the applying section is empty and I don't know how to apply it.

One solution is to compile the project and embbedd the needed fonts and YAML instead of yours in the \data\fonts and \data\themes folder of the project and build a new jar for my project.

Is this the only solution or there is a solution to do  through the API? I use the following code snippet to do the convert :

        Asciidoctor doctor=Asciidoctor.Factory.create();
        Options options=OptionsBuilder.options().headerFooter(true).safe(SafeMode.SAFE).inPlace(true).backend("pdf").get() ;
        doctor.convertFile(new File(asciiFile), options);
 

Is there any solution to pass the YAML and fonts to above API? or I should recompile your project with my YAML and fonts? or there are any other solution?

Thanks

**Comment by mojavelinux** [comment]
I still need to finish the docs :)

In the meantime, to get you going, you need to specify the path to the fonts folder using the `pdf-fontsdir` attribute. The value of this attribute can be any path, relative or absolute. A relative path is resolved relative to current working directory...which isn't correct, but that's how it works currently. Thus, specifying an absolute path is the best strategy atm.

**Comment by mojavelinux** [comment]
> One solution is to compile the project and embbedd the needed fonts and YAML instead of yours in the \data\fonts and \data\themes folder of the project and build a new jar for my project.

I don't recommend going down this path. Let's document what needs to be documented and, if necessary, fix Asciidoctor PDF.

**Comment by mojavelinux** [comment]
Btw, now that I got the gem down to 1.2 MB, I think we can probably just ship a full version of NotoSerif out of the box and avoid this headache. I still think we need a dedicated font for CJK since we likely need a different font for that...but for Cyrillic, we should be able to just use NotoSerif.

**Comment by sohailsed** [comment]
Hi again and thanks

based on your solution, I did the follwing steps

1. Copy my fonts in the d:\myfont folder
2. renamed my fonts to the name your fonts have (for example : notoserif-regular-latin is my non latin regular font)
3. changed my code to :

        Asciidoctor doctor=Asciidoctor.Factory.create();
        Options options=OptionsBuilder.options().headerFooter(true).safe(SafeMode.SAFE).inPlace(true).backend("pdf").get() ;
        options.setOption("pdf-fontsdir","d:\\myfont\\");
        doctor.convertFile(new File(asciiFile), options);

but the problem not fixed and still the not latin font not renedered in the pdf, Am I missed something?

Thx

**Comment by sohailsed** [comment]
Hi again,
 I don't know how to set the path of YAML and use my custom YAML, so I had to change my font name to yours. Is any attributes to set YAML path in a way that I can use my own YAML or better say overriding the default YAML(yours)?

thanks

**Comment by mojavelinux** [comment]
This will be much easier once I finish documenting it.

Until then, I forgot to mention the other required attributes:

* pdf-stylesdir - the path to the directory containing the YAML file
* pdf-style - the name of the YAML file (if the value doesn't end in '.yml', then '-theme.yml' will be appended to the name)
* pdf-fontsdir - the path to the directory containing the fonts

With those three attributes, Asciidoctor PDF should locate your theme and fonts.

**Comment by sohailsed** [comment]
Hi again,

I did the following 

1. copied my fonts to d:\myfonts
2. copied the default-theme.yml to d:\myfonts and change its content to point to my fonts and renamed it to my-theme.yml
3. changed my code to

        Asciidoctor doctor=Asciidoctor.Factory.create();
        Options options=OptionsBuilder.options().headerFooter(true).safe(SafeMode.SAFE).inPlace(true).backend("pdf").get() ;
        options.setOption("pdf-stylesdir","d:\\myfont");
        options.setOption("pdf-style","my-theme.yml");
        options.setOption("pdf-fontsdir","d:\\myfont");
        doctor.convertFile(new File(asciiFile), options);
4. the content of YML is :

font:
  catalog:
    Nazanin:
      normal: Nazanin-Regular.ttf
      bold: Nazanin-Bold.ttf
      italic: Nazanin-Italic.ttf
      bold_italic: Nazanin-BoldItalic.ttf
    # NOTE currently, callout numbers don't work with LiberationMono
    #LiberationMono:
    #  normal: liberationmono-regular-latin.ttf
    #  bold: liberationmono-bold-latin.ttf
    #  italic: liberationmono-italic-latin.ttf
    #  bold_italic: liberationmono-bolditalic-latin.ttf
    Mplus1mn:
      normal: mplus1mn-regular-ascii-conums.ttf
      bold: mplus1mn-bold-ascii.ttf
      italic: mplus1mn-italic-ascii.ttf
      bold_italic: mplus1mn-bolditalic-ascii.ttf
    #Mplus1pMultilingual:
    #  normal: mplus1p-regular-multilingual.ttf
    #FontAwesome:
    #  normal: fontawesome-regular.ttf
  # FIXME using fallbacks breaks use of custom font styles (fixed in Prawn 1.2.1!)
  #fallbacks:
  #  - Mplus1pMultilingual
brand:
  primary_color: 428bca
page:
  background_color: ffffff
  layout: portrait
  # multiply inches by 72 to get pt values
  margin: [0.5 * 72, 0.67 * 72, 0.67 * 72, 0.67 * 72]
  size: Letter
base:
  font_color: 333333
  font_family: Nazanin

------------------------------------

But the problem not fixed. the fonts that embedded in the pdf are still notoserif and mplus1mn and the nazanin(my font) not embedded in the final pdf. Am I miss something?

Thx

**Comment by mojavelinux** [comment]
```
options.setOption("pdf-stylesdir","d:\\myfont");
options.setOption("pdf-style","my-theme.yml");
options.setOption("pdf-fontsdir","d:\\myfont");
```

This is incorrect. You need to specify attributes, not options. Here's what it should be:

```
Attributes attributes = AttributesBuilder.attributes()
  .attribute("pdf-stylesdir", "d:/myfont")
  .attribute("pdf-style", "my")
  .attribute("pdf-fontsdir", "d:/myfont");
options.setAttributes(attributes);
```

**Comment by sohailsed** [comment]
Hi, and thank you.

I did the changes and the non-latin font embedded now in the pdf. But few problem is still on the table:

1. The non-latin characters render left to right char by char and word by word whereas should be render right to left 
2.The characters rendered as separated char whereas should appear as continuous and connected char in the words like situation is in the Arabic language. example: the word سلام which means Hello in English appears م ا ل س which is 4 chars in reverse order and separated chars which is not correct.

Have any ideas to fix?

Thanks

**Comment by mojavelinux** [comment]
Asciidoctor PDF does not yet support RTL writing. (What does work is by chance because it hasn't been tested at all). This is an important, but separate issue. Could you file a new issue so we can start gathering requirements and sort out what is needed to make this work out of the box (or at least with clear instructions)? Until then, I won't know enough to advise you how to fix it. But I certainly want to be able to.

(There's also an issue with RTL writing in Asciidoctor itself. I'm well aware we'll need to address that at some point in the future).

**Comment by sohailsed** [comment]
Hi and thanks for your help.

I opened an issue [here] (https://github.com/asciidoctor/asciidoctor-pdf/issues/175#) 

I hope that we have a solution on this issue as soon as possible for you :)

In the meantime If you had any points to to get I going, let me know,please.

Thanks

**Comment by mojavelinux** [comment]
Resolved original issue with d0ea7b4f21800ae6afe26643cdee0b6210bf19d0.

We'll continue the discussion about RLT language issues in #175. I still have things to learn :)
